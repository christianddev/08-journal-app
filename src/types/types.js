export const types = {
    login: '[Auth] login',
    logout: '[Auth] logout',

    uiSetError: '[UI] setError',
    uiRemoveSetError: '[UI] removeError',

    uiStartLoading: '[UI] Start loading',
    uiFinishLoading: '[UI] Finish loading',

    notesAddNote: '[Notes] Add Note',
    notesActive: '[Notes] Set Active Note',
    notesLoad: '[Notes] Updated Note',
    notesUpdated: '[Notes] Updated Image Url',
    notesDelete: '[Notes] Delete note',
    notesLogoutCleaning: '[Notes] Logout Cleaning',
}