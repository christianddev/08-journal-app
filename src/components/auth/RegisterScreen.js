import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/userForm';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import {serError, removeError }  from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

const RegisterScreen = () => {

  const dispatch = useDispatch();
  const { ui: { msgError } } = useSelector(state => state);

  const [formValues, handleInputChange ]= useForm({
    name: 'userXD',
    email: 'userxd1@email.com',
    password: '123123',
    password2: '123'
  });

  const {name, email, password, password2} = formValues;

  const handleRegister = (event) => {
    event.preventDefault();
    if(isFormValid()) {
      dispatch(startRegisterWithEmailPasswordName(email, password, name));
    }
  }

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(serError('Name is required'));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(serError('Invalid email'));
      return false;
    }else if (password.length < 5) {
      dispatch(serError('Password should be at least 6 characters'));
      return false;
    }else if (password !== password2 || password.length < 5) {
      dispatch(serError('Passwords must match'));
      return false;
    }

    dispatch(removeError());
    return true;
  }

  return (
    <div>
      <h3 className="auth__title mb-5">RegisterScreen</h3>
      <form
        onSubmit={handleRegister}
        className="animate__animated animate__bounce animate__fadeIn animate__faster"
        >

        {
          msgError && ( <div className="auth__alert-error"> {msgError} </div> )
        }

        <input
          type="text"
          placeholder="Name"
          name="name"
          className="auth__input"
          autoComplete="off"
          value={name}
          onChange={handleInputChange}
        />
        <input
          type="text"
          placeholder="Email"
          name="email"
          className="auth__input"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          autoComplete="off"
          value={password}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="Repeat Password"
          name="password2"
          className="auth__input"
          autoComplete="off"
          value={password2}
          onChange={handleInputChange}
        />
        <button
          className="btn btn-primary btn-block mb-5"
          type="submit"
        >
          Register
        </button>
        <Link className="link" to="/auth/login">
          Already register?
        </Link>
      </form>
    </div>
  );
};

export default RegisterScreen;
