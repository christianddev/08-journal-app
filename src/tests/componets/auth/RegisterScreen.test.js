import React from 'react';
import { mount } from "enzyme"
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import RegisterScreen from '../../../components/auth/RegisterScreen';
import '@testing-library/jest-dom';
import { types } from '../../../types/types';

const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const initState = {
  auth:{},
  ui:{
    loading: false,
    msgError: null
  }
}
let store = mockStore(initState)

const wrapper = mount (
  <Provider store={store}>
    <MemoryRouter>
      <RegisterScreen />
    </MemoryRouter>
  </Provider>
)

describe('RegisterScreen.js', () => {

  test('should show', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('should call dispatch action', () => {
    const eventName = {
      target: {
        value: '',
        name: "name"
      }
    };
    wrapper.find('input[name="name"]').simulate("change", eventName);

    wrapper.find('form').simulate('submit',{
      preventDefault(){}
    });
    const actions = store.getActions()
    expect(actions[0]).toEqual({
      type: types.uiSetError,
      payload: 'Name is required'
    })
  })

  test('should show msgError', () => {
    const initState = {
      auth:{},
      ui:{
        loading: false,
        msgError: 'Invalid email'
      }
    }
    const store = mockStore(initState)
    const wrapper = mount (
      <Provider store={store}>
        <MemoryRouter>
          <RegisterScreen />
        </MemoryRouter>
      </Provider>
    )
    expect(wrapper.find('.auth__alert-error').exists()).toBe(true)
    expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initState.ui.msgError)
  })
})
