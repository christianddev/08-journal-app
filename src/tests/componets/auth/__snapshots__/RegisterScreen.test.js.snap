// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterScreen.js should show 1`] = `
<div>
  <h3
    className="auth__title mb-5"
  >
    RegisterScreen
  </h3>
  <form
    className="animate__animated animate__bounce animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Repeat Password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already register?
    </a>
  </form>
</div>
`;
